### ISSUE TITLE
REST API: POST /_replicate can start/stop once, but not restart?

### ISSUE THREAD

**Issue by ahizzle** [couchbase/couchbase-lite-ios#1216]
Per the [API documentation](http://developer.couchbase.com/documentation/mobile/1.2/develop/references/couchbase-lite/rest-api/server/post-replicate/index.html), calling POST /_replicate with a given source/target DB and continuous=true starts a sync task.  

```21:40:21.572‖ Sync: CBLRestPusher[http://localhost:15984/foodb] STARTING ...```

Calling the same endpoint again appears to stop it, per the logging output:

```21:41:42.259‖ Sync: CBLRestPusher[http://localhost:15984/foodb] STOPPING...```

However, all further calls also report STOPPING -- is it possible to re-start a given sync task?  Perhaps we need to use the 'session_id: repl00X' that gets returned as the response to the POST?  If so... my research in the documentation has been futile, so perhaps someone can clue me in.  Thanks! :)

We've also noticed that making the call with continous=false results in the request never returning, though CBlite does appear to be alive (ie: we can see subsequent calls appear in the logs).

--------
* Version:   1.2.0 with fixes for couchdb _changes compatibility cherrypicked
* Client OS: ios simulator
* Server:    couchdb 1.6.1

**Comment by snej** [comment]
Sending another request for continuous replication shouldn't stop it; it should be a no-op. (To stop a replication you add `"cancel":true` to the body.) If it does, that's a bug. Hm.

A POST to start a non-continuous replication does send a response, but not until the replication finishes.

(This is all per CouchDB's replication API.)

**Comment by blairio** [comment]
Hi @snej 

Is it possible to stop a continuous replication using the sessionId (via REST)?

Otherwise, just to clarify - `"cancel":true` should be sent along with all the parameters that were used to start the replication? (including source, target, create_target, continuous, filter, query_params)


Thanks,
=B

**Comment by snej** [comment]
I don't think you can stop a replication from the sessionID.

Yes, keep the other parameters the same and just add `"cancel":true`.

Also, I looked at the source code and noticed that you can add `"async":true` to a non-continuous replication if you want the response sent immediately instead of when it finishes.

**Comment by ahizzle** [comment]
Thanks so much for the quick responses @snej!  Really appreciated.

**Comment by snej** [comment]
I'm still concerned about your statement that a second POST with the same parameters (and no `cancel`) stops the replication. That shouldn't happen, and from looking at the source code of the REST handler, I don't see how it would happen. Could you double-check, please?

**Comment by blairio** [comment]
I have called replicate multiple times with the same arguments and I have not seen replication stop.
However, every once in a while a call to _replicate will hang and not respond. Usually after a bunch of calls to replicate the same DB. Added a timeout and found that after that happened subsequent requests to the listener would also timeout. However we are using fetch from Javascript via React Native, so there could be other things at play.

* Version: 1.2.2 (self built)
* Client OS: iOS simulator 9.3
* Server: couched 1.6.1
* React Native 0.23

**Comment by ahizzle** [comment]
Hi @snej, I think I was just misinterpreting the log output...  I believe the ```21:41:42.259‖ Sync: CBLRestPusher[http://localhost:15984/foodb] STOPPING...``` line is emitted when the sync job has nothing left to sync, but the job is still active.

**Comment by snej** [comment]
OK, closing. @blairio, if you have a reproducible case of the replicator not responding, please file a new issue. Or if you get into this state, try to get a thread dump of the process and see if the Couchbase Lite thread is stuck somewhere.
